core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -172,21 +210,38 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
     }
 
     int lenCopy = len;
+    int maxRetry = Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_READ_RETRY);
+    int retryLeft = maxRetry;
+    boolean retryCurrentWorker = true;
+    Set<WorkerNetAddress> lostWorkers = new HashSet<>();
     while (len > 0) {
       if (pos >= mLength) {
         break;
       }
       long blockId = mStatus.getBlockIds().get(Math.toIntExact(pos / mBlockSize));
       BlockInStream stream = null;
       try {
-        stream = mBlockStore.getInStream(blockId, mOptions);
+        stream = mBlockStore.getInStream(blockId, mOptions, lostWorkers);
         long offset = pos % mBlockSize;
-        int bytesRead =
-            stream.positionedRead(offset, b, off, (int) Math.min(mBlockSize - offset, len));
-        Preconditions.checkState(bytesRead > 0, "No data is read before EOF");
-        pos += bytesRead;
-        off += bytesRead;
-        len -= bytesRead;
+        try {
```

Comment:
we should try to minimize how much needs to be duplicated around each section that might throw an exception due to worker failure. Something like
```java
try {
  // do things that could throw due to worker death
} catch (WorkerDeathExceptions) {
  handleWorkerDeathException();
}
```